Don't pickle LLVM IR for CUDA code libraries

gmarkall opened the pull request:
Pickling an LLVM Module quickly blows up the stack of `reducer_override()` in cloudpickle due to traversal of every function, basic block, instruction, operand, etc. in the module, leading to failures to cache CUDA kernels of any complexity. This PR resolves the issue by pickling LLVM IR as a string instead of as a module.

Additional tests are added to ensure that the unpickled LLVM IR can be used to generate new PTX for devices of previously-unseen compute capability - these tests require a system with multiple Compute Capabilities, so note that they will not be run by gpuCI.

## Effect of the changes

With this patch, the following:

```python
from numba import cuda, njit


def f():
    pass


f_cuda = cuda.jit(cache=True)(f)
f_cpu = njit(cache=True)(f)

print("CUDA call:\n")
f_cuda[1, 1]()

print("\nCPU call:\n")
f_cpu()
```

in conjunction with a small change to Numba to make it explicit what `reducer_override()` is called for:

```diff
diff --git a/numba/cloudpickle/cloudpickle_fast.py b/numba/cloudpickle/cloudpickle_fast.py
index 216450495..d2759126f 100644
--- a/numba/cloudpickle/cloudpickle_fast.py
+++ b/numba/cloudpickle/cloudpickle_fast.py
@@ -641,6 +641,8 @@ class CloudPickler(Pickler):
               reducers, such as Exceptions. See
               https://github.com/cloudpipe/cloudpickle/issues/248
             """
+
+            print(f"reducer_override() for {type(obj)}")
             if sys.version_info[:2] < (3, 7) and _is_parametrized_type_hint(obj):  # noqa  # pragma: no branch
                 return (
                     _create_parametrized_type_hint,
```

produces:

```
CUDA call:

/home/gmarkall/numbadev/numba/numba/cuda/dispatcher.py:502: NumbaPerformanceWarning: Grid size 1 will likely result in GPU under-utilization due to low occupancy.
  warn(NumbaPerformanceWarning(msg))
reducer_override() for <class 'numba.core.typing.templates.Signature'>
reducer_override() for <class 'type'>
reducer_override() for <class 'numba.core.types.misc.NoneType'>
reducer_override() for <class 'function'>
reducer_override() for <class 'function'>
reducer_override() for <class 'numba.core.types.abstract._TypeMetaclass'>
reducer_override() for <class 'type'>
reducer_override() for <class 'numba.cuda.codegen.CUDACodeLibrary'>
reducer_override() for <class 'function'>
reducer_override() for <class 'numba.core.serialize._CustomPickled'>
reducer_override() for <class 'method'>
reducer_override() for <class 'type'>
reducer_override() for <class 'function'>
reducer_override() for <class 'function'>
reducer_override() for <class 'type'>
reducer_override() for <class 'code'>
reducer_override() for <class 'type'>
reducer_override() for <class 'function'>
reducer_override() for <class 'type'>
reducer_override() for <class 'abc.ABCMeta'>
reducer_override() for <class 'numba.core.callconv._MinimalCallHelper'>
reducer_override() for <class 'type'>

CPU call:

reducer_override() for <class 'numba.core.funcdesc.PythonFunctionDescriptor'>
reducer_override() for <class 'type'>
reducer_override() for <class 'numba.core.types.misc.NoneType'>
reducer_override() for <class 'function'>
reducer_override() for <class 'function'>
reducer_override() for <class 'numba.core.types.abstract._TypeMetaclass'>
reducer_override() for <class 'type'>
reducer_override() for <class 'numba.core.environment.Environment'>
reducer_override() for <class 'function'>
reducer_override() for <class 'numba.core.typing.templates.Signature'>
reducer_override() for <class 'type'>
```

There is a little more to pickle with the CUDA target than the CPU target due to some CUDA-target-specific extras, but the items pickled are generally similar. Without this patch, `reducer_override()` has to be called much more as the entire module is pickled. A small excerpt from the output for the CUDA call without this patch:

```
reducer_override() for <class 'llvmlite.ir.module.Module'>
reducer_override() for <class 'type'>
reducer_override() for <class 'llvmlite.ir.context.Context'>
reducer_override() for <class 'type'>
reducer_override() for <class 'llvmlite.ir._utils.NameScope'>
reducer_override() for <class 'type'>
reducer_override() for <class 'collections.defaultdict'>
reducer_override() for <class 'type'>
reducer_override() for <class 'type'>
reducer_override() for <class 'llvmlite.ir._utils.NameScope'>
reducer_override() for <class 'collections.defaultdict'>
reducer_override() for <class 'collections.OrderedDict'>
reducer_override() for <class 'type'>
reducer_override() for <class 'llvmlite.ir.values.Function'>
reducer_override() for <class 'type'>
reducer_override() for <class 'llvmlite.ir.types.PointerType'>
reducer_override() for <class 'type'>
reducer_override() for <class 'llvmlite.ir.types.FunctionType'>
reducer_override() for <class 'type'>
reducer_override() for <class 'llvmlite.ir.types.IntType'>
reducer_override() for <class 'type'>
reducer_override() for <class 'llvmlite.ir.types.PointerType'>
reducer_override() for <class 'llvmlite.ir.types.PointerType'>
reducer_override() for <class 'llvmlite.ir.types.IntType'>
... <and much more>
```

which quickly blows up the stack for anything more complex, e.g. the example given in #8373.

Fixes #8373.

## To do

- [x] Tidy up additional tests
- [x] Add test of failing to pickle an unfinalized `CUDACodeLibrary`.
- [x] Test these changes locally with cuDF

gmarkall commented:
gpuci run tests

gmarkall commented:
gpuci run tests

gmarkall commented:
gpuci run tests

gmarkall commented:
gpuci run tests

sklam commented on numba/cuda/tests/cudapy/test_caching.py:
```diff
@@ -190,6 +203,15 @@ def test_non_writable_pycache(self):
         self._test_pycache_fallback()
 
+    def test_cannot_cache_linking_libraries(self):
+        link = os.path.join(Path(__file__).parent.parent,
+                            'cudadrv', 'data', 'jitlink.ptx')
```
FYI, could have used `/` (slash operator) instead of join; e.g. `Path(__file__).parent.parent / 'cudadrv' / 'data' / 'jitlink.ptx'`

gmarkall replied:
Huh, interesting - I didn't know that!

sklam commented on numba/cuda/codegen.py:
```diff
@@ -337,6 +341,8 @@ def _rebuild(cls, codegen, name, entry_name, module, linking_libraries,
         instance._max_registers = max_registers
         instance._nvvm_options = nvvm_options
 
+        instance._finalized = True
```
good catch

gmarkall commented:
gpuci run tests

gmarkall commented:
@sklam Many thanks for the initial review. I believe this is now ready for a "proper" review. Notes on the changes since you looked yesterday:

- I've added the test that pickling an unfinalized `CUDACodeLibrary` fails.
- I noticed that construction of the unpickled `CUDACodeLibrary` was completely wrong - it used `__new__()` for no reason then called the wrong `__init__()` method. I'd noticed before that there was some weirdness in that it appeared `CUDACodeLibrary().__init__()` was not being called to initialize its members, and this was the cause. It seems much more straightforward to use the usual call syntax to create the unpickled object, and I note that this is what other `_rebuiild()` methods do too.
- I came to the conclusion that I shouldn't split up the multi-CC cache test because that would need a bit more duplication - instead I've added some explanatory comments which I hope make it easier to understand the test.
- In addition to passing on gpuCI, this passes locally on my machine which has multiple compute capabilities. 
- I've run this with cuDF and all cuDF tests pass. As a quick experiment I also added `cache=True` to the JIT decorators in [`test_extension_compilation.py`](https://github.com/rapidsai/cudf/blob/branch-22.10/python/cudf/cudf/tests/test_extension_compilation.py) and it knocked about 10% off the runtime (170 vs 184 seconds) - the tests do do a lot of work besides just compiling kernels so I think this is a significant speedup in Numba jitting time. Modifying the rest of cuDF to actually cache UDFs is more involved so I haven't tried that yet.

sklam reviewed (approved):
Thanks for the patch.
